common/pkg/libartifact/store/reference.go

```diff
@@ -0,0 +1,100 @@
+//go:build !remote
+
+package store
+
+import (
+	"context"
+
+	"go.podman.io/image/v5/docker/reference"
+)
+
+type ArtifactReference struct {
+	reference.Named
+}
+
+// NewArtifactReference is a theoretical reference to an artifact.  It needs to be
+// a fully qualified oci reference except for tag, where we add
+// "latest" as the tag if tag is empty.  Valid references:
+//
+// quay.io/podman/machine-os:latest
+// quay.io/podman/machine-os
+// quay.io/podman/machine-os@sha256:916ede4b2b9012f91f63100f8ba82d07ed81bf8a55d23c1503285a22a9759a1e
+//
+// Note: Partial sha references and digests (IDs) are not allowed.
+func NewArtifactReference(input string) (ArtifactReference, error) {
+	ar := ArtifactReference{}
+	named, err := stringToNamed(input)
+	if err != nil {
+		return ArtifactReference{}, err
+	}
+	ar.Named = named
+	return ar, nil
+}
+
+func (ar ArtifactReference) IsDigested() bool {
+	_, isDigested := ar.Named.(reference.Digested)
+	return isDigested
+}
+
+type ArtifactStoreReference struct {
+	ArtifactFromStore *Artifact
+	IsDigested        bool
+	Ref               reference.Named
+}
+
+// NewArtifactStorageReference refers to an object already in the artifact store.  It
+// can be a name or a full or partial digest.  Conveniently, it also embeds the artifact
+// as part of its return.
+func NewArtifactStorageReference(nameOrDigest string, as *ArtifactStore) (ArtifactStoreReference, error) {
+	lookupInput := nameOrDigest
+	asf := ArtifactStoreReference{}
+	al, err := as.getArtifacts(context.Background(), nil)
+	if err != nil {
+		return ArtifactStoreReference{}, err
+	}
+
+	// Try to parse as a valid OCI reference
+	named, parseErr := stringToNamed(nameOrDigest)
+	if parseErr == nil {
+		lookupInput = named.String()
+	}
+
+	// Lookup in the store
+	a, isDigest, err := al.getByNameOrDigest(lookupInput)
+	if err != nil {
+		return ArtifactStoreReference{}, err
+	}
+
+	// If parsing failed, parse the artifact's name instead
+	if parseErr != nil {
+		fqName, err := a.GetName()
```

Comment:
[MULTI], see elsewhere.

… this assigns _some_ of possibly several names matching the digest to the returned value.

And then the name is used as the primary lookup key, e.g. in `ArtifactStore.Remove`.

That does _something_, but it’s not clear that it’s what the user wanted.

Again, can we entirely delete the “ID” lookup case? Or at least have it fail if there are multiple matching objects? Or would we want to have it fail only in _some_ situations? `delete digestValue` is a problem; `inspect digestValue` might be fine.

---

… but this also applies to digest lookups: `pull repo@theDigest; pull repo:tagPointingToThatDigest; pull repo:tagAlsoPointingToThatDigest` creates 3 separate named objects in the store, but `getByNameOrDigest` will only return one (~arbitrarily-chosen) name.